community/values/src/main/java/org/neo4j/values/virtual/NodeValue.java

```diff
@@ -68,4 +63,107 @@ public String toString()
         return format( "(%d)", id );
     }
 
+    static class DirectNodeValue extends NodeValue
+    {
+        private final TextArray labels;
+        private final MapValue properties;
+
+        DirectNodeValue( long id, TextArray labels, MapValue properties )
+        {
+            super( id );
+            assert labels != null;
+            assert properties != null;
+            this.labels = labels;
+            this.properties = properties;
+        }
+
+        public TextArray labels()
+        {
+            return labels;
+        }
+
+        public MapValue properties()
+        {
+            return properties;
+        }
+    }
+
+    public static class NodeProxyWrappingNodeValue extends NodeValue
+    {
+        private final Node node;
+        private volatile TextArray labels;
+        private volatile MapValue properties;
+
+        NodeProxyWrappingNodeValue( Node node )
+        {
+            super( node.getId() );
+            this.node = node;
+        }
+
+        public Node nodeProxy()
+        {
+            return node;
+        }
+
+        @Override
+        public <E extends Exception> void writeTo( AnyValueWriter<E> writer ) throws E
+        {
+            TextArray l;
+            MapValue p;
+            try
+            {
+                l = labels();
+                p = properties();
+            }
+            catch ( NotFoundException e )
+            {
+                l = Values.stringArray();
+                p = VirtualValues.EMPTY_MAP;
+
+            }
+            writer.writeNode( node.getId(), l, p );
+        }
+
+        @Override
+        public TextArray labels()
+        {
+            TextArray l = labels;
+            if ( l == null )
+            {
+                synchronized ( this )
```

Comment:
Aha, yes I see. What I actually was wondering about was the *concurrency-safiness*?